scanpy/external/tl/_sam.py

```diff
@@ -0,0 +1,216 @@
+"""\
+Run the Self-Assembling Manifold algorithm
+"""
+from typing import Optional, Union
+
+from anndata import AnnData
+
+from ... import logging as logg
+
+
+def sam(
+    adata: AnnData,
+    max_iter: int = 10,
+    num_norm_avg: int = 50,
+    k: int = 20,
+    distance: str = 'correlation',
+    standardization: Optional[str] = 'Normalizer',
+    weight_pcs: bool = True,
+    npcs: Optional[int] = None,
+    n_genes: Optional[int] = None,
+    projection: Optional[str] = 'umap',
+    inplace: bool = True,
+    verbose: bool = True,
+) -> Optional[AnnData]:
+    """Self-Assembling Manifolds single-cell RNA sequencing analysis tool.
+
+    SAM iteratively rescales the input gene expression matrix to emphasize
+    genes that are spatially variable along the intrinsic manifold of the data.
+    It outputs the gene weights, nearest neighbor matrix, and a 2D projection.
+
+    The AnnData input should contain unstandardized, non-negative values.
+    Preferably, the data should be log-normalized and no genes should be filtered out.
+
+
+    Parameters
+    ----------
+
+    k - int, optional, default 20
+        The number of nearest neighbors to identify for each cell.
+
+    distance : string, optional, default 'correlation'
+        The distance metric to use when identifying nearest neighbors.
+        Can be any of the distance metrics supported by sklearn's 'pdist'.
+
+    max_iter - int, optional, default 10
```

Comment:
```suggestion
    max_iter
```

Please remove all the types and defaults from the docstring, they’re added automatically

Reply:
Done!